### PR TITLE
[backend] Decay rules settings (#2859)

### PR DESCRIPTION
**Pull request by SouadHadjiat**
### Proposed changes

* Create a new module for decay rules
* Initialize built-in decay rules (3 decay rules saved at init => update counters in tests + 3)
* Save decay rules in cache
* Apply saved decay rules (from cache) on indicators
* List and display decay rules in settings
* Add a form to create decay rules
* generic tsx use for both chart (on settings and on indicator overview)
* View and edit decay rule => + chart to visualize decay rule
* Delete decay rules from settings
* Reset decay computations when an indicator has an upsert on score or a score update (except if score = base score).
* Add a simple visualization in graph for past decay data (when decay has been reset because of upsert or update).
* Remove feature flag

image

image

### Related issues

* #2859 

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: opencti-platform/opencti-front/lang/front/en.json [hunk 1]
```diff
@@ -2328,8 +2328,8 @@
   "Reference knowledge update": "Reference knowledge update",
   "Enable commit messages": "Enable commit messages",
   "Commit message when updating data from the UI.": "Commit message when updating data from the UI.",
-  "To use this feature, your platform administrator must enable the subscription manager in the config.": "To use this feature, your platform administrator must enable the subscription manager in the config.",
-  "To use this feature, your platform administrator must enable the according manager in the config.": "To use this feature, your platform administrator must enable the according manager in the config.",
+  "To use this feature, your platform administrator must enable the subscription manager in the config.": "To use this feature, your platform administrator must enable the subscription manager in configuration.",
+  "To use this feature, your platform administrator must enable the according manager in the config.": "To use this feature, your platform administrator must enable the according manager in configuration.",
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/lang/front/en.json:2332]: In all other language "configuration" seems to be whole word.

File: opencti-platform/opencti-graphql/tests/03-streams/00-Raw/raw-test.js [hunk 2]
```diff
@@ -23,13 +23,13 @@ describe('Raw streams tests', () => {
       expect(createEventsByTypes.label.length).toBe(15);
       expect(createEventsByTypes.identity.length).toBe(30);
       expect(createEventsByTypes.relationship.length).toBe(126);
-      expect(createEventsByTypes.indicator.length).toBe(31);
+      expect(createEventsByTypes.indicator.length).toBe(33);
```
Comment (aHenryJard) [inline opencti-platform/opencti-graphql/tests/03-streams/00-Raw/raw-test.js:26]: DecayRule integration test creates 2 indicators, so this is legit to me (same for others +2).

File: opencti-platform/opencti-front/src/components/i18n.jsx [hunk 3]
```diff
@@ -213,6 +218,33 @@ export const useFormatter = () => {
...
+   */
+  const minuteHourDate = (date) => {
+    if (isNone(date)) {
+      return '-';
+    }
+    return moment(date).format('DD MMM yyyy HH:mm A');
+  };
+
+  /**
+   * Relaive from now in word. Like "2 months ago", "in 3 minutes"...
```
Comment (aHenryJard) [inline opencti-platform/opencti-front/src/components/i18n.jsx]:
```suggestion
   * Relative from now in word. Like "2 months ago", "in 3 minutes"...
```

File: opencti-platform/opencti-front/src/private/components/observations/indicators/DecayDialog.tsx [hunk 4]
```diff
@@ -12,85 +12,104 @@ import { SxProps } from '@mui/material';
...
+      label: t_i18n('Stability threshold'),
+      style: { color: theme.palette.text.primary },
+      score: history.score,
+      updated_at: getDateAsTextFor(history),
+    };
   };
 
-  const decayFullHistory: LabelledDecayHistory[] = [];
+  const labelledHistoryList: LabelledDecayHistory[] = [];
   decayHistory.map((history) => (
```
Comment (jpkha) [inline opencti-platform/opencti-front/src/private/components/observations/indicators/DecayDialog.tsx:91]: const labelledHistoryList = [...decayHistory, ...decayLivePoints].map(history => getDisplayFor(history))

File: opencti-platform/opencti-front/src/private/components/settings/Root.jsx [hunk 5]
```diff
@@ -215,6 +217,15 @@ const Root = () => {
...
           />
+          <BoundaryRoute
+            exact
+            path="/dashboard/settings/customization/decay"
+            component={Decay}
+          />
+          <BoundaryRoute
+            path="/dashboard/settings/customization/decay/:decayRuleId"
+            component={RootDecay}
+          />
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/Root.jsx]: In my opinion here we should have only one route `/dashboard/settings/customization/decay` that points to `decay/Root.tsx` and in `decay/Root.tsx` define 2 relative paths `""` and `":decayId"`

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRule.tsx [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import DecayChart, { DecayHistory } from '@components/settings/decay/DecayChart';
+import DecayRuleEdition from './DecayRuleEdition';
+import DecayRulePopover from './DecayRulePopover';
+import ExpandableMarkdown from '../../../../components/ExpandableMarkdown';
+import { useFormatter } from '../../../../components/i18n';
+import ItemBoolean from '../../../../components/ItemBoolean';
+import CustomizationMenu from '../CustomizationMenu';
+import { DecayRule_decayRule$key } from './__generated__/DecayRule_decayRule.graphql';
+
+const useStyles = makeStyles(() => ({
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRule.tsx:21]: `makeStyles` is deprecated. Try to put style directly in JSX using `style={{...}}` for html elements or `sx={{...}}` for MUI elements

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRuleCreation.tsx [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+import TextField from '../../../../components/TextField';
+import type { Theme } from '../../../../components/Theme';
+import SwitchField from '../../../../components/SwitchField';
+import { fieldSpacingContainerStyle } from '../../../../utils/field';
+import { insertNode } from '../../../../utils/store';
+import useAuth from '../../../../utils/hooks/useAuth';
+import { handleErrorInForm } from '../../../../relay/environment';
+import { DecayRulesLinesPaginationQuery$variables } from './__generated__/DecayRulesLinesPaginationQuery.graphql';
+
+const useStyles = makeStyles<Theme>((theme) => ({
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRuleCreation.tsx:30]: `makeStyles` is deprecated. Try to put style directly in JSX using `style={{...}}` for html elements or `sx={{...}}` for MUI elements

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRuleEdition.tsx [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+import MarkdownField from '../../../../components/MarkdownField';
+import TextField from '../../../../components/TextField';
+import type { Theme } from '../../../../components/Theme';
+import SwitchField from '../../../../components/SwitchField';
+import { fieldSpacingContainerStyle } from '../../../../utils/field';
+import useAuth from '../../../../utils/hooks/useAuth';
+import { handleError } from '../../../../relay/environment';
+import { DecayRule_decayRule$data } from './__generated__/DecayRule_decayRule.graphql';
+
+const useStyles = makeStyles<Theme>((theme) => ({
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRuleEdition.tsx:28]: `makeStyles` is deprecated. Try to put style directly in JSX using `style={{...}}` for html elements or `sx={{...}}` for MUI elements

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRulePopover.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+import MoreVert from '@mui/icons-material/MoreVert';
+import makeStyles from '@mui/styles/makeStyles';
+import { useNavigate } from 'react-router-dom-v5-compat';
+import { DecayRule_decayRule$data } from './__generated__/DecayRule_decayRule.graphql';
+import { decayRuleEditionMutation } from './DecayRuleEdition';
+import Transition from '../../../../components/Transition';
+import { useFormatter } from '../../../../components/i18n';
+import { handleError } from '../../../../relay/environment';
+
+const useStyles = makeStyles(() => ({
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRulePopover.tsx:20]: `makeStyles` is deprecated. Try to put style directly in JSX using `style={{...}}` for html elements or `sx={{...}}` for MUI elements

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRules.tsx [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import ListLines from '../../../../components/list_lines/ListLines';
+import type { Theme } from '../../../../components/Theme';
+import { DecayRulesLinesPaginationQuery, DecayRulesLinesPaginationQuery$variables } from './__generated__/DecayRulesLinesPaginationQuery.graphql';
+import { DecayRulesLine_node$data } from './__generated__/DecayRulesLine_node.graphql';
+import useAuth from '../../../../utils/hooks/useAuth';
+import { INDICATOR_DECAY_MANAGER } from '../../../../utils/platformModulesHelper';
+
+const LOCAL_STORAGE_KEY = 'view-decay-rules';
+
+const useStyles = makeStyles<Theme>(() => ({
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRules.tsx:22]: `makeStyles` is deprecated. Try to put style directly in JSX using `style={{...}}` for html elements or `sx={{...}}` for MUI elements

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRulesLine.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+import ListItemIcon from '@mui/material/ListItemIcon';
+import ListItemText from '@mui/material/ListItemText';
+import makeStyles from '@mui/styles/makeStyles';
+import Skeleton from '@mui/material/Skeleton';
+import ItemIcon from '../../../../components/ItemIcon';
+import { DataColumns } from '../../../../components/list_lines';
+import type { Theme } from '../../../../components/Theme';
+import { DecayRulesLine_node$key } from './__generated__/DecayRulesLine_node.graphql';
+
+const useStyles = makeStyles<Theme>((theme) => ({
```
Comment (lndrtrbn) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRulesLine.tsx:16]: `makeStyles` is deprecated. Try to put style directly in JSX using `style={{...}}` for html elements or `sx={{...}}` for MUI elements

File: opencti-platform/opencti-front/src/private/components/settings/decay/DecayRule.tsx [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    margin: '6px 0 0 0',
+    padding: '15px',
+    borderRadius: 6,
+  },
+  popover: {
+    float: 'left',
+    marginTop: '-13px',
+  },
+  title: {
+    float: 'left',
```
Comment (jpkha) [inline opencti-platform/opencti-front/src/private/components/settings/decay/DecayRule.tsx:45]: Check if we really need to use `float` css property

**Review by aHenryJard — APPROVED** [review]

**Pull request merged**